src/rules/noKeywordNamedVariablesRule.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2014 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as Lint from "../lint";
+import * as ts from "typescript";
+
+const BAD_NAMES = [ "any", "Number", "number", "String", "string", "Boolean", "boolean", "undefined" ];
```

Comment:
@JKillian started with this to kick off discussion.

Arguably this list could include any ambient variable from `lib.d.ts`, like [`Number`](https://github.com/Microsoft/TypeScript/blob/55218d64a58a9964229daa0a38bdfb0a6be717e5/lib/lib.d.ts#L519), [`String`](https://github.com/Microsoft/TypeScript/blob/55218d64a58a9964229daa0a38bdfb0a6be717e5/lib/lib.d.ts#L519) and [`Boolean`](https://github.com/Microsoft/TypeScript/blob/55218d64a58a9964229daa0a38bdfb0a6be717e5/lib/lib.d.ts#L457) 

The idea behind this rule is to increase the likelihood of writing code that is unambiguously understood by as many as possible. You could easily chuck `RegExp`, `Array`and other core interfaces that have real class equivalents in there too... 

But then you could probably well argue it goes too far to include [`event`](https://github.com/Microsoft/TypeScript/blob/55218d64a58a9964229daa0a38bdfb0a6be717e5/lib/lib.d.ts#L16738) for example.